backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/predicate/impl/AbstractLuceneSearchNestedPredicateBuilder.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.search.predicate.impl;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import org.hibernate.search.util.common.AssertionFailure;
+
+import org.apache.lucene.search.Query;
+
+public abstract class AbstractLuceneSearchNestedPredicateBuilder extends AbstractLuceneSearchPredicateBuilder {
+
+	private final List<String> nestedPathHierarchy;
+
+	public AbstractLuceneSearchNestedPredicateBuilder(List<String> nestedPathHierarchy) {
+		this.nestedPathHierarchy = nestedPathHierarchy;
+	}
+
+	@Override
+	public final Query build(LuceneSearchPredicateContext context) {
+		List<String> nestedSteps = implicitNestedSteps( context, nestedPathHierarchy );
+		return applyImplicitNestedSteps( nestedSteps, context, super::build );
+	}
+
+	public static List<String> implicitNestedSteps(LuceneSearchPredicateContext context, List<String> nestedPathHierarchy) {
```

Comment:
I'm not sure I get the value of this method, and in particular the value of copying the list of nested paths.
You managed to do without copying the list in the Elasticsearch backend, by simply looping over *all* paths and skipping the ones that are not relevant. Can't we do the same here?
Deep down, it boils down to simply using code similar to the Elasticsearch backend.

Reply:
You're right, we can just use the context to skip any explicit already performed nested step.